etc/docker/kernel-image-puller/kernel_image_puller.py

```diff
@@ -10,286 +10,310 @@
+        self.runtime_endpoint = None
+        self.default_container_registry = None
+        self.log = kip_logger
+        self.worker_queue = None
+        self.threads = []
+        self.pulled_images = []
+        self.num_pullers = None
+        self.num_retries = None
+        self.policy = None
+        self.validate_cert = None
+        self.load_static_env_values()
+
+    def load_static_env_values(self):
+        self.num_pullers = int(os.getenv("KIP_NUM_PULLERS", "2"))
+        self.num_retries = int(os.getenv("KIP_NUM_RETRIES", "3"))
+        self.policy = os.getenv("KIP_PULL_POLICY", KernelImagePuller.POLICY_IF_NOT_PRESENT)
+        self.default_container_registry = os.getenv("KIP_DEFAULT_CONTAINER_REGISTRY", "docker.io")
+        self.runtime_endpoint = os.getenv(
+            "KIP_CRI_ENDPOINT", "unix:///run/containerd/containerd.sock"
         )
-
-    if kspecs is None:
-        return False
-
-    # Locate the configured images within the kernelspecs and add to set for duplicate management
-    images = set()
-    for key in kspecs.keys():
-        metadata = kspecs.get(key).get("spec").get("metadata")
-        if metadata is not None:
-            process_proxy = metadata.get("process_proxy")
-            if process_proxy is not None:
-                config = process_proxy.get("config")
-                if config is not None:
-                    image_name = config.get("image_name")
-                    if image_name is not None:
-                        images.add(image_name)
-                    executor_image_name = config.get("executor_image_name")
-                    if executor_image_name is not None:
-                        images.add(executor_image_name)
-
-    if not images:
-        return False
-
-    # Add the image names to the name queue to be pulled
-    for image_name in images:
-        name_queue.put_nowait(image_name)
-    return True
-
-
-def pull_image(image_name):
-    """Pulls the image.
-
-    If the policy is `IfNotPresent` the set of pulled image names is
-    checked and, if present, the method returns.  Otherwise, the pull attempt is made
-    and the set of pulled images is updated, when successful.
-    """
-    if policy == POLICY_IF_NOT_PRESENT:
-        if image_name in pulled_images:
-            # Image has been pulled, but make sure it still exists.  If it doesn't exist
-            # let this drop through to actual pull
-            logger.info(
-                f"Image '{image_name}' already pulled and policy is '{policy}'.  Checking existence."
+        self.container_runtime = self.get_container_runtime()
+        self.gateway_host = os.getenv("KIP_GATEWAY_HOST", "http://localhost:18888")
+        # Add authentication token support to KIP
+        self.auth_token = os.getenv("KIP_AUTH_TOKEN", None)
+        self.interval = int(os.getenv("KIP_INTERVAL", "300"))
+        self.validate_cert = os.getenv("JUPYTER_GATEWAY_VALIDATE_CERT", "False").lower() == "true"
```

Comment:
There is one issue here: How do we make `JUPYTER_GATEWAY_VALIDATE_CERT` env available to KIP pod unless we pass it via the daemonset spec.  One option is to completely replace `JUPYTER_GATEWAY_VALIDATE_CERT` -> `KIP_VALIDATE_CERT`. 

The reason I did used `JUPYTER_GATEWAY_VALIDATE_CERT` was because its already used by gatewayclient and the certificate under consideration belongs to GATEWAY server and not KIP.

Reply:
True, the defaulting I suggested doesn't work.  So _this_ `JUPYTER_GATEWAY_VALIDATE_CERT` has nothing to do with the env used back on the _client_ server.  I think we should just define `KIP_VALIDATE_CERT` completely independently since these are completely separate processes and users.

Reply:
got it. made all the changes.